### DISCUSSION TITLE
Galax

### DISCUSSION THREAD

**Discussion by prashjet** [dynamics-of-stellar-systems/tnt#3] [General]
- email from Nathaniel Starkman 26/06/26
   - We have several major pushes planned for `galax`. The most relevant right now is that we've nearly finished a refactor of `coordinax`, our coordinate library, which will simplify a lot of `galax` internals and make it faster and easier to use.
   - On documentation: Adrian is working on adapting the `gala` tutorials to `galax`, which should lower the barrier to entry considerably.
   - We're also working with [`odisseo`](https://github.com/vepe99/Odisseo) and [`xstream`](https://arxiv.org/abs/2508.02666), two other libraries that are building on `galax` under the hood.
   - chat planned for next week:
      - ask re continuity

**Comment by prashjet** [comment]
Noted from chats with Nathaniel Starkman 7/7/26 & 16/7/26

- changes to galax
   - several, on near-term (this week) and long-term
   - API "relatively stable but with but some changes"
   - conclusion: no harm in starting now and trying to adapt when changes arrive / freezing at some early galax version if necessary

- ecosystem:
   - upstream: 
      - quax: multiple dispatch
      - unxt: units
      - coordinax: coordinates and reference frames
   - downstream:
      - Odisseo, Xstream, seems like there is significant interaction between communities
      - tnt...? would be good to be part of a larger ecosystem!

- a lot of the complexity of galax arises since it handles units
   - based on [quax](https://github.com/nstarman/quax) for multiple dispatch with JAX
   - multiple dispatch = different methods for different input types
   - relation to galax: if you provide unit-ful quantities, these are carried through the function. Also possible to provide plain JAX arrays
   - there is [significant performance hit](https://coordinax.readthedocs.io/en/latest/guides/perf.html) from carrying units through traced JAX functions, but galax is (mostly) written in a way you can avoid this

- direct acceleration calculations
   - currently, accelerations come from auto-diff on the potnential, with a performance hit
   - within scope to add direct force calculations
   - in practice, add a ``_gradient`` method to the potential objects, to over-ride the default gradient from auto-diff
   - add tests when impleneting these

- TriaxialGaussianMGE
   - triaxial NFW density already implemented, with Chandrasekhar formula
   - for performance: look into dropping the order of integration from 50 → 10,
   - inner/outer approximation regimes:
      - i.e. use taylor approx for small r, dipole term for large r
      - can accomplish this with ``jax.lax.select_n`` i.e. equivalent of `jnp.where` with > 2 conditions
      - Triaxial Gaussian --> Triaxial MGE: may need to bring the summation inside the integral (check if this holds for different q' per MGE cmp.)

- other
   - hypothesis for property based testing
      - you write tests which should pass for all inputs in whatever range you describe, and let Hypothesis randomly choose which of those inputs to check - including edge cases you might not have thought about
   - Claude Fable successes
